jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartInputStreamParser.java

```diff
@@ -0,0 +1,794 @@
+        public void parsedField(String key, String value)
+        {
+            // Add to headers and mark if one of these fields. //
+            headers.put(key.toLowerCase(Locale.ENGLISH), value);
+            if (key.equalsIgnoreCase("content-disposition"))
+                contentDisposition=value;
+            else if (key.equalsIgnoreCase("content-type"))
+                contentType = value;    
+
+        }
+
+        @Override
+        public boolean headerComplete() {
+        
+            try {
+                
+                // Extract content-disposition
+                boolean form_data=false;
+                if(contentDisposition==null)
+                {
+                    throw new IOException("Missing content-disposition");
+                }
+    
+                QuotedStringTokenizer tok=new QuotedStringTokenizer(contentDisposition,";", false, true);
+                String name=null;
+                String filename=null;
+                while(tok.hasMoreTokens())
+                {
+                    String t=tok.nextToken().trim();
+                    String tl=t.toLowerCase(Locale.ENGLISH);
+                    if(t.startsWith("form-data"))
+                        form_data=true;
+                    else if(tl.startsWith("name="))
+                        name=value(t);
+                    else if(tl.startsWith("filename="))
+                        filename=filenameValue(t);
+                }
+    
+                // Check disposition
+                if(!form_data)
+                {
+                    return false;
+                }
+                //It is valid for reset and submit buttons to have an empty name.
+                //If no name is supplied, the browser skips sending the info for that field.
+                //However, if you supply the empty string as the name, the browser sends the
+                //field, with name as the empty string. So, only continue this loop if we
+                //have not yet seen a name field.
+                if(name==null)
+                {
+                    return false;
+                }
+
+                //create the new part
+                _part = new MultiPart(name, filename);
+                _part.setHeaders(headers);
+                _part.setContentType(contentType);
+                _parts.add(name, _part);
+            }
+            catch (Exception e)
+            {
+                _err = e;
+                return true;
+            }
+            
+            return false;
+        }
+        
+        @Override
+        public boolean content(ByteBuffer buffer, boolean last)
+        {
+            if (BufferUtil.hasContent(buffer)) 
+            {
+                try
+                {
+                    //write the content data to the part
+                    _part.open();
+                    _part.write(buffer.array(),buffer.arrayOffset()+buffer.position(),buffer.remaining());
+                    _part.close();
```

Comment:
This looks wrong.  I would have expected the part to be opened in headerComplete, then multiple calls to content would each call write and then finally the part to be closed in messageComplete.

Reply:
messageComplete is only called when the entire form is parsed and not after each individual parts content section. So I could call `_part.open()` in headerComplete, and `_part.close()` if last, in the content method.